### PR TITLE
fix(blosc2): split large inputs across SChunk chunks (#68)

### LINKED ISSUE #68: tensogram (core): blosc2 codec fails on arrays > ~2 GB
### What happened?

Encoding a large array (e.g., [288, 1332, 1072] float64 ≈ 1.6 GB) with blosc2 codec fails with:

`EncodingError: compression error: blosc2 error: MaxBufsizeExceeded`

**What should happen:** Either automatic chunking within the codec (`blosc2` supports internal blocking), or a clear error message with guidance to use `zstd` or `lz4` for large arrays, or a documented max buffer size in the compression options.

`zstd` codec works fine on the same data.

### What are the steps to reproduce the bug?

```python
import tensogram
import numpy as np
import tempfile, os
tmp = tempfile.mkdtemp()
tgm_path = os.path.join(tmp, 'large_array.tgm')
# ~1.6 GB array (288 * 1332 * 1072 * 8 bytes)
shape = [288, 1332, 1072]
size_gb = np.prod(shape) * 8 / (1024**3)
print(f'Array: shape={shape}, dtype=float64, size={size_gb:.2f} GB')
arr = np.zeros(shape, dtype='float64')  # zeros to avoid memory pressure
meta = {'version': 2}
objects = [
    ({'type': 'ntensor', 'shape': shape, 'dtype': 'float64', 'compression': 'blosc2'},
     arr),
]
try:
    with tensogram.TensogramFile.create(tgm_path) as f:
        f.append(meta, objects)
    print('SUCCESS (unexpected)')
except Exception as e:
    print(f'CRASH: {type(e).__name__}: {e}')
# Show that zstd works on the same data
tgm_path2 = os.path.join(tmp, 'large_array_zstd.tgm')
objects2 = [
    ({'type': 'ntensor', 'shape': shape, 'dtype': 'float64', 'compression': 'zstd'},
     arr),
]
try:
    with tensogram.TensogramFile.create(tgm_path2) as f:
        f.append(meta, objects2)
    size_mb = os.path.getsize(tgm_path2) / (1024*1024)
    print(f'zstd works: {size_mb:.1f} MB')
except Exception as e:
    print(f'zstd also failed: {e}')
" 2>&1
Array: shape=[288, 1332, 1072], dtype=float64, size=3.06 GB
CRASH: ValueError: EncodingError: compression error: blosc2 error: MaxBufsizeExceeded
zstd works: 0.1 MB
```

### Version

0.16.1

### Platform (OS and architecture)

Fedora43 x86_64

### Relevant log output

```shell
CRASH: ValueError: EncodingError: compression error: blosc2 error: MaxBufsizeExceeded
```

### Accompanying data

_No response_

### Organisation

_No response_

### PR DESCRIPTION
**Pull request by sametd**
## Description

Fixes [#68](https://github.com/ecmwf/tensogram/issues/68): the `blosc2` codec crashes with `EncodingError: compression error: blosc2 error: MaxBufsizeExceeded` on arrays larger than ~2 GiB. The reporter's example — `[288, 1332, 1072]` float64 ≈ 3.06 GiB, a routine ECMWF weather-grid shape — fails today; `zstd` works on the same inputs.

### Root cause

C-Blosc2 has a hard per-call compression limit of `BLOSC2_MAX_BUFFERSIZE = INT_MAX - 32 ≈ 2.147 GiB`, defined in `c-blosc2/include/blosc2.h` as a structural feature of the on-the-wire chunk format (32-bit signed length header). Tensogram's `Blosc2Compressor::compress()` was calling `SChunk::new()` + `schunk.append(data)` with the entire tensor buffer in a single call, so every blosc2 chunk hit the C limit on multi-GiB inputs. `zstd` escapes the limit because its frame format uses 64-bit sizes; blosc2 cannot lift it at the codec level.

A second, subtler bug surfaced while fixing the first: the Rust `blosc2` crate v0.2.2 computes `SChunk::items_num()` as `num_chunks * chunksize / typesize` (`blosc2-0.2.2/src/chunk/schunk.rs:466-468`), which **over-reports** whenever the final chunk is shorter than `chunksize` — the common case once we split a large input across multiple chunks. The pre-fix `decompress()` called `schunk.items(0..items_num())`, which would then ask the C layer for phantom items past the real end of the data and fail. This was caught by Oracle review before implementation and is the reason `decompress()` also needed changes, not just `compress()`.

### Fix

1. **Encode**: split the input into SChunk chunks of up to `DEFAULT_BLOSC2_CHUNK_BYTES` (256 MiB, matching python-blosc2's upper cap). Each chunk's `append()` call stays well below the 2 GiB limit. One trailing chunk may be short; c-blosc2's frame format explicitly tolerates a single short tail chunk.
2. **Decode**: iterate chunks explicitly via `schunk.get_chunk(i).decompress()`, which uses each chunk's own recorded `nbytes` (`blosc2-0.2.2/src/chunk/chunk.rs:141`) instead of the broken `items_num()` path.
3. **Defensive**: clamp effective `chunk_bytes` at `BLOSC2_MAX_BUFFERSIZE` inside `compress()` so direct struct construction with a misconfigured value (the field is public) can't re-trigger #68.

For inputs smaller than `chunk_bytes` the single-append fast path preserves **byte-for-byte-identical output** to the pre-fix codepath, so existing golden/determinism tests continue to pass unchanged.

### Commits

| SHA | Type | Description |
|---|---|---|
| `a15deb8` | `fix` | Core #68 fix: multi-chunk compress + per-chunk decompress + pipeline wire-up + 6 regression tests |
| `24490c8` | `fix` | Defensive clamp at `BLOSC2_MAX_BUFFERSIZE` against public-field misconfiguration |
| `dd50d3f` | `test` | Edge-case coverage: non-typesize-aligned tail (reachable from `simple_packing` with non-byte-aligned `bits_per_value`) |
| `0c5fa05` | `docs` | `IDEAS.md` entry for the `block_offsets` follow-up spun out during review |

### Testing

- **+8 new blosc2 unit tests** (8 pre-existing → 16 total). Covers: multi-chunk short tail, multi-chunk exact multiple, range decode across a chunk boundary, small-input single-chunk byte-identity, input exactly equal to `chunk_bytes`, typesize alignment, non-typesize-aligned tail, `chunk_bytes > BLOSC2_MAX_BUFFERSIZE` clamp. The critical regression test is `blosc2_multi_chunk_round_trip_short_tail`, which would have failed against the original `decompress()` and catches the `items_num()` miscount.
- **Green quad on the full workspace**: `cargo fmt --check`, `cargo test -p tensogram-encodings` (all pass), `cargo test -p tensogram --lib` (405/405 pass — includes the Wire v3 / mask-aware tests from `main`), `cargo clippy --workspace --all-targets --all-features -- -D warnings` (clean), LSP diagnostics (none).
- **Determinism preserved**: `blosc2_nthreads_zero_is_deterministic_across_runs` continues to pass. For inputs ≤ `chunk_bytes` the single-append path produces byte-identical output to the pre-fix codepath.

### Trade-offs

| Axis | Direction | Detail |
|---|---|---|
| Arrays > 2 GiB actually encode | ✅ | The whole point of this PR |
| Compression ratio on > 256 MiB inputs | tiny ↓ | Each chunk boundary resets the backend compressor's context. Typically < 1–2% for scientific float data with shuffle+zstd; not benchmarked (noted as a follow-up in the file) |
| Output byte overhead | tiny ↓ | ~32 B per extra chunk + small SChunk index. 3 GiB array → 12 chunks → ~400 B extra |
| Peak memory during encode | ↑ | Bounded by `chunk_bytes` (256 MiB) instead of the full input |
| `decompress_range` on large arrays | ↑ | Only decompresses chunks overlapping the requested range |
| Byte-level output for arrays ≤ 256 MiB | = | Identical to pre-fix (single-append fast path) |
| Wire format / message format | = | SChunk output is a standard blosc2 super-chunk, readable by python-blosc2, c-blosc2, h5py+hdf5plugin, PyTables. No tensogram format version bump |
| Backward decode compatibility | = | Old single-chunk files decode fine with new code; new multi-chunk files decode fine with the existing chunk-iteration logic |

### Review process

This fix went through an **Oracle architecture review before implementation** (which caught the `items_num()` dealbreaker in the original plan and required the decompress path to also be rewritten), followed by **Pass 2 simplification**, **Pass 3 hardening** (the defensive clamp + non-aligned tail test), **Pass 4 polish**, and a **final Oracle green-light review**. The commit split mirrors those quality gates.

### Follow-ups (out of scope, in `plans/IDEAS.md`)

- **blosc2 `block_offsets`** — expose per-chunk byte offsets via the existing `CompressResult.block_offsets` slot (parallel to szip's existing `"szip_block_offsets"` key), enabling byte-ranged remote reads from S3/GCS and amortising the `SChunk::from_buffer` header-parse cost on hot `decompress_range` loops. c-blosc2's `blosc2_frame_get_offsets()` is a stable public API; `blosc2_rs_sys` surfaces it. Deferred until a concrete use case or microbenchmark motivates the wire-format addition.
- **`simple_packing + blosc2` with non-byte-aligned `bits_per_value`** — pre-existing alignment concern, orthogonal to #68. Now has a regression test in the suite, but any bug is a separate issue.
- **Cross-codec `Vec::with_capacity` allocation hardening** — project-wide concern, not specific to blosc2.

### Upstream-dependency watch-out

If `c-blosc2` ever changes `BLOSC2_MAX_OVERHEAD` from 32 bytes, the mirrored private const at `blosc2.rs:42` must be kept in sync. Currently pinned to `blosc2 = "0.2"`, so the risk window is the next crate upgrade.

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)

## Checklist

- [x] I have read the [Contributing Guide](CONTRIBUTING.md)
- [x] My code follows the project's style guidelines (\`cargo fmt\`, \`ruff format\`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All new and existing tests pass (\`cargo test --workspace\`, \`pytest\`)
- [x] I have updated documentation where necessary
- [x] I have added examples if this introduces new public API

## Contributor Licence Agreement

By submitting this pull request, I confirm that my contribution is made under
the terms of the [Apache License 2.0](LICENSE) and that I have the right to
submit it under this licence. I agree to the terms of the
[ECMWF Contributor Licence Agreement](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md).

Docs Preview
https://sites.ecmwf.int/docs/tensogram/pull-requests/PR-69

### CODE REVIEW

File: rust/tensogram-encodings/src/compression/blosc2.rs [hunk 1]
```diff
@@ -97,21 +159,42 @@ impl Compressor for Blosc2Compressor {
...
+        let num_chunks = schunk.num_chunks();
+        if num_chunks == 0 {
             return Ok(Vec::new());
         }
-        schunk.items(0..num_items).map_err(map_err)
+
+        let mut out = Vec::with_capacity(expected_size);
+        for idx in 0..num_chunks {
+            let chunk = schunk.get_chunk(idx).map_err(map_err)?;
+            let bytes = chunk.decompress().map_err(map_err)?;
```
[the remark this answers is not in the document]
Comment (sametd) [reply rust/tensogram-encodings/src/compression/blosc2.rs]: Good catch — applied in commit 1c31807 ("fix(blosc2): don't preallocate decompress buffer from caller metadata").

I went with a simpler version of your suggested patch:

```rust
let mut out: Vec<u8> = Vec::new();
for idx in 0..num_chunks {
    let chunk = schunk.get_chunk(idx).map_err(map_err)?;
    let bytes = chunk.decompress().map_err(map_err)?;
    out.try_reserve(bytes.len()).map_err(|e| {
        CompressionError::Blosc2(format!(
            "failed to reserve {} bytes for decompressed chunk {idx}: {e}",
            bytes.len(),
        ))
    })?;
    out.extend_from_slice(&bytes);
}
```

Differences from your suggestion:

- **No outer `try_reserve` based on `expected_size`.** Since the whole point is that `expected_size` is attacker-derived, I prefer to drop trust on it entirely rather than carry it into a capped reserve. `Vec::new()` + amortised per-chunk `try_reserve` is simpler and doesn't require picking a cap policy.
- **`chunk.nbytes()` as the per-reserve source.** The size we reserve per chunk comes from the SChunk frame trailer (parsed by `blosc2-rs`), not from the pipeline's descriptor-derived estimate. An attacker could still craft a SChunk with a pathological `nbytes`, but that's the underlying `blosc2` crate's allocation path, not ours.
- **Renamed `expected_size` → `_expected_size`** and added a security-related comment explaining why we drop it, so a future "unused parameter" cleanup doesn't silently plug it back into a preallocation.

I also did a quick audit of the rest of the codecs (result in `plans/TODO.md` under Code Quality → "cross-codec `expected_size` preallocation hardening"). Summary:

| Codec | Status |
|---|---|
| **blosc2** (this PR) | Fixed — uses `try_reserve` + fallible path |
| **szip** (`libaec.rs:150`) | Still vulnerable — `vec![0u8; expected_size]` actually *worse* (zero-initialises) |
| **szip-pure** (`tensogram-szip/src/decoder.rs:88`) | Still vulnerable — `Vec::with_capacity(total_samples)` |
| zstd / zstd-pure / lz4 / zfp / sz3 / rle / roaring | Safe — ignore `expected_size` or validate it against decompressed output |

The cross-codec fix needs a coordinated change (either an upper bound in `estimate_decompressed_size` at the pipeline layer, or `try_reserve` in each remaining codec), which Oracle recommended keeping out of a focused bug-fix PR. TODO.md entry `d400f4c` tracks the work.

Thanks for the catch — this genuinely did introduce an attack vector my earlier review missed.

File: rust/tensogram-encodings/src/compression/blosc2.rs [hunk 2]
```diff
@@ -51,6 +78,19 @@ pub struct Blosc2Compressor {
...
+    ///
+    /// Production callers set this to [`DEFAULT_BLOSC2_CHUNK_BYTES`]
+    /// (256 MiB) via the pipeline constructor; tests may set it smaller
+    /// to exercise the multi-chunk path without large allocations. Values
+    /// smaller than `typesize` (including `0`) are rounded up to exactly
+    /// one `typesize` element per chunk.
+    ///
+    /// [`compress`]: Compressor::compress
+    pub chunk_bytes: usize,
 }
```
[the remark this answers is not in the document]
Comment (sametd) [reply rust/tensogram-encodings/src/compression/blosc2.rs:82]: Good catch, applied in `3fe9655` ("refactor(blosc2): move chunk_bytes from pub field to private test seam").

You're right that adding a public field is a semver break — I had been leaning on the precedent that `ZstdCompressor::nb_workers` and `Blosc2Compressor::nthreads` were both added as pub fields in v0.13.0 under a minor version bump, so I figured 0.17.0 was the right target. But Oracle pushed back on that: unlike `nthreads`/`nb_workers` which are real caller-facing tuning knobs, `chunk_bytes` was **test scaffolding leaking into the public API** — it existed only so unit tests could exercise the multi-chunk path with 16 KiB buffers instead of allocating the 256 MiB production default per test.

The cleaner fix is to keep the test seam private:

```rust
impl Blosc2Compressor {
    /// Private test seam: production callers reach this via
    /// `Compressor::compress` with `DEFAULT_BLOSC2_CHUNK_BYTES`.
    fn compress_with_chunk_bytes(
        &self,
        data: &[u8],
        requested_chunk_bytes: usize,
    ) -> Result<CompressResult, CompressionError> { ... }
}

impl Compressor for Blosc2Compressor {
    fn compress(&self, data: &[u8]) -> Result<CompressResult, CompressionError> {
        self.compress_with_chunk_bytes(data, DEFAULT_BLOSC2_CHUNK_BYTES)
    }
}
```

Tests call `compressor.compress_with_chunk_bytes(&data, 4096)` directly; the public surface of `Blosc2Compressor` is now exactly what it was on `origin/main` before this PR. Ships as 0.16.2 (patch) with zero public API delta.

Also dropped `DEFAULT_BLOSC2_CHUNK_BYTES` from the `compression::mod` re-export for the same reason — pipeline doesn't need it anymore, and keeping it there would lock us into supporting a constant we never promised.

Net: `git diff origin/main..HEAD -- '*.rs' | grep "^+pub"` returns nothing new in this PR.

Thanks for catching this — it was a real API-hygiene issue, not just a semver formality.

File: rust/tensogram-encodings/src/compression/blosc2.rs [hunk 3]
```diff
@@ -88,7 +128,29 @@ impl Compressor for Blosc2Compressor {
...
-        schunk.append(data).map_err(map_err)?;
+
+        // Floor `chunk_bytes` to a multiple of `typesize` so non-tail
+        // appends stay aligned for blosc2's shuffle filter, clamp upward
+        // to at least one `typesize` to guarantee forward progress, and
+        // clamp downward to `BLOSC2_MAX_BUFFERSIZE` so the multi-append
+        // path stays legal even if a caller misconfigures `chunk_bytes`.
+        let ts = self.typesize.max(1);
+        let capped = self.chunk_bytes.min(BLOSC2_MAX_BUFFERSIZE);
+        let chunk_bytes = (capped / ts).max(1) * ts;
```
[the remark this answers is not in the document]
Comment (sametd) [reply rust/tensogram-encodings/src/compression/blosc2.rs]: Good find, applied in the same commit `3fe9655`.

Your arithmetic is correct: with `typesize > capped`, the expression `(capped / ts).max(1) * ts` evaluates to `ts`, which can (in theory) exceed `BLOSC2_MAX_BUFFERSIZE`. In practice the blosc2 crate's `CParams::typesize()` validates typesize to `[1, 255]` (`blosc2-0.2.2/src/params.rs:126-134`), so `build_cparams()` would error out before `compress_with_chunk_bytes()` ever reaches the clamp — the bug is unreachable today. But the inline comment claims the clamp "stays legal ... regardless of how the caller ... arrived here" and the code didn't actually honour that guarantee. Making the code match its documentation is free.

Applied the suggested single-character fix inside the new private helper:

```rust
let chunk_bytes = ((capped / ts).max(1) * ts).min(BLOSC2_MAX_BUFFERSIZE);
```

I folded this into the refactor commit rather than splitting it out because both changes touch the same 5-line region of `compress_with_chunk_bytes` and the clamp makes more sense inside the new private helper anyway. The commit message on `3fe9655` names the refactor as the primary driver; this reply is the paper trail for Comment 2.

No regression test for this specific path — the only way to reach `ts > BLOSC2_MAX_BUFFERSIZE` is to bypass `build_cparams()`'s typesize validation, which would require a deeper refactor than this PR warrants. The defensive clamp is pure "match the documented invariant" with no behaviour change for any reachable input.

File: rust/tensogram-encodings/src/compression/blosc2.rs [hunk 4]
```diff
@@ -79,39 +107,113 @@ impl Blosc2Compressor {
...
-}
 
-impl Compressor for Blosc2Compressor {
-    fn compress(&self, data: &[u8]) -> Result<CompressResult, CompressionError> {
+    /// Compress `data` into an SChunk whose per-chunk byte size is at most
+    /// `requested_chunk_bytes`.  Private test seam so that unit tests can
+    /// exercise the multi-chunk path with small buffers without having to
+    /// allocate [`DEFAULT_BLOSC2_CHUNK_BYTES`] worth of scratch memory.
+    /// Production callers reach this via [`Compressor::compress`], which
+    /// passes `DEFAULT_BLOSC2_CHUNK_BYTES`.
```
[the remark this answers is not in the document]
Comment (sametd) [reply rust/tensogram-encodings/src/compression/blosc2.rs]: Good catch — applied in `515b18a` ("docs: clarify chunk-size rounding + disambiguate PR/issue refs").

You're right that the "at most `requested_chunk_bytes`" claim is wrong in the `requested_chunk_bytes < typesize` corner — `(capped / ts).max(1) * ts` evaluates to `ts` in that case, which exceeds the request.

Went with option (a) from your comment (fix the doc, not the code) because rounding UP to one `typesize` is the right semantics: you can't have an SChunk chunk that holds less than one element of the typed data. Erroring out on `requested < typesize` would trade one uncomfortable edge case for a noisier one, and silently truncating data would be wrong.

The new docstring:
- Describes the floor-to-typesize + clamp-to-max semantics explicitly
- Calls out the round-up edge case in plain language
- Notes that production callers never reach it (`DEFAULT_BLOSC2_CHUNK_BYTES` = 256 MiB ≫ max typesize = 255 B per blosc2's validation)

The last point is important for a future reader: without it, someone might see the "effective size can exceed requested" line and think it's a real bug that needs a test. Documenting that it's unreachable in practice prevents that confusion.

No regression test for the edge case — the only way to reach it is to call the private helper with `requested_chunk_bytes < typesize` directly in a test. I didn't add one because (a) the behavior is a side-effect of integer math, not a documented contract, and (b) the existing multi-chunk tests already exercise the path with sensible requests. Let me know if you'd prefer an explicit regression test anyway.

File: plans/TODO.md [hunk 5]
```diff
@@ -296,6 +296,34 @@ For speculative ideas, see `IDEAS.md`.
...
+    - `Compressor::decompress` receives `expected_size` from
+      `estimate_decompressed_size(config)`
+      (`rust/tensogram-encodings/src/pipeline.rs:787`), which multiplies
+      `config.num_values * dtype_byte_width` with no upper-bound check.
+      `config.num_values` is attacker-controllable via the tensor-shape
+      CBOR of a malformed `.tgm` file, so a pathological value drives
+      an infallible allocation that can abort the process (DoS).
+    - blosc2 was fixed in-PR for #69 by dropping `expected_size` trust
+      entirely and using `Vec::new()` + per-chunk `try_reserve` (size
+      comes from the codec's own frame trailer, not the descriptor).
```
[the remark this answers is not in the document]
Comment (sametd) [reply plans/TODO.md]: Good catch — applied in the same commit `515b18a`.

You're right that `#69` standalone is ambiguous (GitHub uses the same `#N` syntax for issues and PRs). Updated the TODO entry:

```diff
-    - blosc2 was fixed in-PR for #69 by dropping `expected_size` trust
-      entirely and using `Vec::new()` + per-chunk `try_reserve` (size
-      comes from the codec's own frame trailer, not the descriptor).
+    - blosc2 was fixed in PR #69 (which closes issue #68) by dropping
+      `expected_size` trust entirely and using `Vec::new()` + per-chunk
+      `try_reserve` (size comes from the codec's own frame trailer, not
+      the descriptor).
```

Also caught two more ambiguous refs in the same commit while I was at it:
- `TODO.md` line 326: "Spun out of the Copilot review on #69" → "on PR #69"
- `IDEAS.md` line 93: "the #68 multi-chunk fix review" → "the issue #68 multi-chunk fix review (PR #69)"

The principle is: once this PR merges, the TODO/IDEAS entries should survive as standalone artefacts without requiring the reader to reconstruct the `#68` vs `#69` context from git history. "PR #69 (issue #68)" makes that crystal clear even for someone reading the plan files cold.

File: rust/tensogram-encodings/src/compression/blosc2.rs [hunk 6]
```diff
@@ -323,4 +443,202 @@ mod tests {
...
+    #[test]
+    fn blosc2_chunk_bytes_above_max_buffersize_is_clamped() {
+        let data: Vec<u8> = (0..8192).map(|i| (i % 251) as u8).collect();
+        let compressor = small_chunk_compressor();
+        let compressed = compressor
+            .compress_with_chunk_bytes(&data, BLOSC2_MAX_BUFFERSIZE + (1 << 30))
+            .unwrap()
+            .data;
+        let decompressed = compressor.decompress(&compressed, data.len()).unwrap();
+        assert_eq!(decompressed, data);
```
[the remark this answers is not in the document]
Comment (sametd) [reply rust/tensogram-encodings/src/compression/blosc2.rs]: Nice catch, applied in `eb7ea7b` ("test(blosc2): unit-test effective_chunk_bytes clamp directly").

You're right that the original test was weak: with `data.len() = 8192` the fast-path condition `data.len() <= chunk_bytes` always held, so the clamp was never actually load-bearing for the test's assertion. Verified this locally by temporarily removing the `.min(BLOSC2_MAX_BUFFERSIZE)` and the test still passed.

Did **both** of the options you suggested rather than just one:

1. **Extracted the clamp into a free function** `effective_chunk_bytes(requested, typesize) -> usize` that can be unit-tested in isolation, no multi-GiB allocations needed. The helper lives next to `blosc2_nthreads` and follows the same "small pure function with a docstring explaining the invariants" pattern.

2. **Added four property-style unit tests** that exercise the real clamp invariants:
   - `effective_chunk_bytes_never_exceeds_blosc2_max`: matrix of `typesize ∈ {1..255}` × `request ∈ {0, BLOSC2_MAX_BUFFERSIZE ± {0, 1, 2^30}, usize::MAX}` asserts every result is `≤ BLOSC2_MAX_BUFFERSIZE`. **This is the test that would have caught the regression that shipped this PR's first draft** — no `.min()` at the end of the arithmetic and this matrix would fail on every `usize::MAX` cell.
   - `effective_chunk_bytes_is_typesize_aligned`: verifies `result % typesize == 0` and `result >= typesize` for all combinations.
   - `effective_chunk_bytes_rounds_up_below_typesize`: pins the `requested < typesize` edge case.
   - `effective_chunk_bytes_zero_typesize_is_defensive`: verifies the `typesize == 0` branch is well-defined.

3. **Renamed the original integration test** to `blosc2_oversized_chunk_bytes_request_round_trips` and rewrote its docstring to be honest: "shallow integration test — only proves `compress()` doesn't crash on an oversized request; the actual clamp-value property is verified by the unit tests on `effective_chunk_bytes` below."

Side bonus: the six-line inline comment in `compress_with_chunk_bytes` collapsed into the helper's docstring, so the clamp invariants now live in one documented place rather than two.

`cargo test -p tensogram-encodings --lib compression::blosc2::` is 20/20 passing (was 16/16 — the 4 new unit tests plus the renamed integration test).

Thanks — this kind of "named for X but tests Y" smell is exactly the kind of thing that rots over years. Good catch.
